// @fellow/coffeelint2/lib/rules/spacing_after_comma.js
(function() {
  var SpacingAfterComma;

  module.exports = SpacingAfterComma = (function() {
    class SpacingAfterComma {
      constructor() {
        this.inRegex = false;
      }

      lintToken(token, tokenApi) {
        var type;
        [type] = token;
        if (type === 'REGEX_START') {
          this.inRegex = true;
          return;
        }
        if (type === 'REGEX_END') {
          this.inRegex = false;
          return;
        }
        if (!(token.spaced || token.newLine || this.isGenerated(token, tokenApi) || this.isRegexFlag(token, tokenApi))) {
          return {token};
        }
      }

      // Coffeescript does some code generation when using CSX syntax, and it adds
      // brackets & commas that are not marked as generated. The only way to check
      // these is to see if the comma has the same column number as the last token.
      isGenerated(token, tokenApi) {
        var offset, pos, prevPos, prevToken;
        if (token.generated) {
          return true;
        }
        offset = -1;
        prevToken = tokenApi.peek(offset);
        while (prevToken.generated) {
          offset -= 1;
          prevToken = tokenApi.peek(offset);
        }
        pos = token[2];
        prevPos = prevToken[2];
        if (pos.first_line === prevPos.first_line && pos.first_column === prevPos.first_column) {
          return true;
        }
        return false;
      }

      // When generating a regex (///${whatever}///i) CoffeeScript generates tokens
      // for RegEx(whatever, "i") but doesn't bother to mark that comma as
      // generated or spaced. Looking 3 tokens ahead skips the STRING and CALL_END
      isRegexFlag(token, tokenApi) {
        var maybeEnd;
        if (!this.inRegex) {
          return false;
        }
        maybeEnd = tokenApi.peek(3);
        return (maybeEnd != null ? maybeEnd[0] : void 0) === 'REGEX_END';
      }

    };

    SpacingAfterComma.prototype.rule = {
      name: 'spacing_after_comma',
      level: 'ignore',
      message: 'a space is required after commas',
      description: 'This rule checks to make sure you have a space after commas.'
    };

    SpacingAfterComma.prototype.tokens = [',', 'REGEX_START', 'REGEX_END'];

    return SpacingAfterComma;

  }).call(this);

}).call(this);
